### PR TITLE
`set_block` with `symmetry_`

### PR DESCRIPTION
**Pull request by JonathonMisiewicz**
## Description
See below.

## User API & Changelog headlines
- [x] `Matrix::set_block` can be used on matrices that are not totally symmetric.

## Dev notes & details
- [x] `Matrix::set_block` can be used on matrices that are not totally symmetric.
- [x] `Matrix::set_block` now raises an explicit error when the symmetry of the target and block matrices disagree.

## Checklist
- [x]  New test working

## Status
- [x] Ready for review
- [x] Ready for merge

### CODE REVIEW

File: psi4/src/psi4/libmints/matrix.cc [hunk 1]
```diff
@@ -673,17 +673,20 @@ void Matrix::set_block(const Slice &rows, const Slice &cols, const Matrix& block
...
+    auto block_cols = cols.end() - cols.begin();
     for (int h = 0; h < nirrep_; h++) {
         int max_p = block_rows[h];
-        int max_q = block_cols[h];
+        int max_q = block_cols[h ^ symmetry_];
         for (int p = 0; p < max_p; p++) {
             for (int q = 0; q < max_q; q++) {
                 double value = block.get(h, p, q);
-                set(h, p + rows_begin[h], q + cols_begin[h], value);
+                set(h, p + rows_begin[h], q + cols_begin[h ^ symmetry_], value);
```
Comment (loriab) [inline psi4/src/psi4/libmints/matrix.cc:689]: So previous behavior was the matrix would be set to wrong values if it wasn't symmetric?

Comment (JonathonMisiewicz) [reply psi4/src/psi4/libmints/matrix.cc:689]: That or segfaulting.

**Review by loriab — APPROVED** [review]

File: psi4/src/psi4/libmints/matrix.cc (hunk 1 above)
Comment (psi-rking) [reply psi4/src/psi4/libmints/matrix.cc:689]: Our EOM-CC codes have LOTS of the "h^irrep" motif in them.  Before the story is lost to history, we should press TDC and CDS if they remember the source of the innovation.  I think it went back beyond Curt and Ed, so maybe back to psi1!?

File: tests/pytests/test_matrix.py [hunk 2]
```diff
@@ -225,3 +226,23 @@ def test_doublets(adl, adr, Ga, bdl, bdr, Gb, at, bt):
...
+            mat.set(1, k, j, 10 + i)
+            if i not in {0, 2}:
+                control_mat.set(0, j, k, i)
+                control_mat.set(1, k, j, 10 + i)
+            i += 1
+    new_dim = Dimension([2, 1])
+    new_slice = Slice(Dimension([0, 0]), new_dim)
+    new_mat = Matrix("Name", new_dim, new_dim, 1)
+    mat.set_block(new_slice, new_slice, new_mat)
+    assert psi4.compare_matrices(mat, control_mat)
```
Comment (psi-rking) [inline tests/pytests/test_matrix.py:248]: After staring for a few minutes and finally figuring out what this test is testing, LGTM !

**Review by psi-rking — APPROVED** [review]

File: psi4/src/psi4/libmints/matrix.cc (hunk 1 above)
Comment (JonathonMisiewicz) [reply psi4/src/psi4/libmints/matrix.cc:689]: That traces back to the MOLECULE code. See Section 4 of [this report that TDC has saved](https://drive.google.com/file/d/1VtVpwvaleR1FF1hm2LchS2SQZ8_RHsJA/view). Molecule - A Program System for Non-Empirical Calculation of the Electronic Structure of Molecules II. Integral Section. University of Stockholm Institute of Physics Report 74 - 29, December 1974.

Hat tip to TDC for remembering this.

**Comment by psi-rking** [comment]
Cool!  I noticed "The extension to higher symmetries does not seem
promising, however."  That looks pretty solid given that it's 2022 and we
still don't do it!

**Review by maxscheurer — APPROVED** [review]

**Review by lothian — APPROVED** [review]

**Pull request merged**